### PR TITLE
Rename IFakeOptionsBuilder to IFakeOptions

### LINKED ISSUE #520: Rename `IFakeOptionsBuilder` to `IFakeOptions`
Extracted from https://github.com/FakeItEasy/FakeItEasy/issues/461#issuecomment-118212168, since

1. I think it's a good move anyhow, and 
2. If we do this, it'll affect #461, and may simplify our discussion of that issue

----

The more I look at things, the more I think that what users see as "fake options builders" could really just be "fake options" to them. That would alleviate _some_ of the naming pressure (for the API) in #461, not just in the discoverable interface we're talking about there, but when using `A.Fake<T>(Action<IFakeOptionsBuilder<T>>)` too.

If we made the change, the internal `FakeOptions` would collide with the class we'd use to implement the public `IFakeOptions`.
We could:

1. rename `FakeOptions` to something like `FakeCreationOptions`, or
2. collapse `FakeOptionsBuilder` and `FakeOptions`, as I'm not sure the latter is pulling its weight). If we did this we _could_ still implement two interfaces, to separate the "build up the options" actions from the "just get me the data so I can make a proxy" operations.

### PR DESCRIPTION
**Pull request by blairconrad**
Fixes #520.

Contains steps 1–4 from #526, plus the obligatory "cleanup" commit.

connects to #520

### CODE REVIEW

**Comment by jimmyheaddon** [comment]
For reference, pull request [#536](https://github.com/FakeItEasy/FakeItEasy/pull/536) will need merging with this one at some point.

**Comment by blairconrad** [comment]
Yup, @jimmyheaddon. My plan was to, assuming the other owners have no problem with #536, merge that one first and rebase my PR.

**Comment by blairconrad** [comment]
Rebased! I think we're ready for review.

**Comment by blairconrad** [comment]
Unassigning. Since we now have a veritable army of owners, in general there's no benefit in assigning a PR to anyone in particular.

File: Source/FakeItEasy/Core/FakeManager.AutoFakePropertyRule.cs [hunk 1]
```diff
@@ -11,7 +11,7 @@ public partial class FakeManager
         private class AutoFakePropertyRule
             : IFakeObjectCallRule
         {
-            public FakeManager FakeManager { get; set; }
+            public FakeManager FakeManager { private get; set; }
```
Comment (thomaslevesque) [inline Source/FakeItEasy/Core/FakeManager.AutoFakePropertyRule.cs]: The write-only property is a bit weird... Wouldn't it be better to make it private and initialize it from the constructor?

**Comment by thomaslevesque** [comment]
I just reviewed the code. There are many changes, but most of them are trivial, so I don't have much to say (just one minor remark).

(I did learn something along the way: I didn't know about the `[Fake]` and `[UnderTest]` attributes. That's great stuff! I was using an auto-mocking container to achieve the same result, but the attribute-based approached is much better.)

**Comment by blairconrad** [comment]
Thanks. I enjoyed your remark and will look at tonight.

The attributes aren't bad. I've never used them on my own. Sometimes I think it's a "cool feature that maybe belongs outside the core package". But I think things like that about lots of stuff.
I'm anti-feature, in general.

**Comment by adamralph** [comment]
@blairconrad for my understanding, can you please explain the rationale behind the first commit:

> renaming fakeConfigurationActions to proxyConfigurationActions

**Comment by adamralph** [comment]
Oh, and also the second commit:

> renaming FakeOptions to ProxyOptions

**Comment by thomaslevesque** [comment]
@adamralph, it's explained in #526

**Comment by adamralph** [comment]
@thomaslevesque thanks! Memory like a sieve!

BTW - @blairconrad, I love the separation of commits here, it makes this so much easier to review (and only heightens my embarrassment for not doing so sooner).

File: Source/FakeItEasy/Core/FakeManager.AutoFakePropertyRule.cs (hunk 1 above)
Comment (adamralph) [inline Source/FakeItEasy/Core/FakeManager.AutoFakePropertyRule.cs]: Agreed, but I think we can deal with this in a separate PR. This change only highlights the design problem by minimising the accessibility to what is needed, and thus highlights the design problem.

**Comment by blairconrad** [comment]
Thanks, gentlemen. I was a little occupied last night and today or would've gotten into the conversation.
@thomaslevesque, nice explaining about #526. I probably would've forgotten as well.
@adamralph, if you like those commits, wait until you see how I broke the next change up into 22 simple commits!
And don't be embarrassed.

**Pull request merged**